resolwe_bio/processes/reads_processing/trimgalore.py

```diff
@@ -0,0 +1,451 @@
+                "trimming, this option will simply hard-trim sequences "
+                "to bp from the 3' end. This is incompatible with "
+                "other hard trimming options.",
+                required=False,
+            )
+            trim_3 = IntegerField(
+                label="Hard trim sequences from 5' end",
+                description="Instead of performing adapter-/quality "
+                "trimming, this option will simply hard-trim sequences "
+                "to bp from the 5' end. This is incompatible with "
+                "other hard trimming options.",
+                required=False,
+            )
+
+        adapter_trim = GroupField(AdapterTrimming, label="Adapter trimming")
+        quality_trim = GroupField(QualityTrimming, label="Quality trimming")
+        hard_trim = GroupField(HardTrimming, label="Hard trimming")
+
+    class Output:
+        """Output fields."""
+
+        fastq = ListField(FileField(), label="Remaining mate 1 reads")
+        fastq2 = ListField(FileField(), label="Remaining mate 2 reads")
+        report = FileField(label="Trim galore report", required=False)
+        fastqc_url = ListField(
+            FileHtmlField(), label="Mate 1 quality control with FastQC"
+        )
+        fastqc_url2 = ListField(
+            FileHtmlField(), label="Mate 2 quality control with FastQC"
+        )
+        fastqc_archive = ListField(FileField(), label="Download mate 1 FastQC archive")
+        fastqc_archive2 = ListField(FileField(), label="Download mate 2 FastQC archive")
+
+    def run(self, inputs, outputs):
+        """Run analysis."""
+
+        if inputs.adapter_trim.adapter and inputs.adapter_trim.adapter_file_1:
+            self.error(
+                "Mate 1 adapters should be either a sequence or a file, but not both."
+            )
+
+        if inputs.adapter_trim.adapter_2 and inputs.adapter_trim.adapter_file_2:
+            self.error(
+                "Mate 2 adapters should be either a sequence or a file, but not both."
+            )
+
+        if inputs.hard_trim.trim_5 and inputs.hard_trim.trim_3:
+            self.error("Only one type of hard trimming can be performed at once.")
+
+        mate1_path = os.path.basename(inputs.reads.fastq[0].path)
+        assert mate1_path.endswith(".fastq.gz")
+        name_mate1 = mate1_path[:-9]
+        mate2_path = os.path.basename(inputs.reads.fastq2[0].path)
+        assert mate2_path.endswith(".fastq.gz")
+        name_mate2 = mate2_path[:-9]
+
+        merged_r1 = "input_reads_mate1.fastq.gz"
+        merged_r1_name = merged_r1[:-9]
+        merged_r2 = "input_reads_mate2.fastq.gz"
+        merged_r2_name = merged_r2[:-9]
+        (Cmd["cat"][[reads.path for reads in inputs.reads.fastq]] > merged_r1)()
+        (Cmd["cat"][[reads.path for reads in inputs.reads.fastq2]] > merged_r2)()
+
+        params = [
+            "--paired",
+            "--cores",
+            2,  # Actual core usage is 9
```

Comment:
Why hardcode this?

Reply:
The default is to only run on a single core, this overrides it.

Reply:
But why did't you use the value from the process requirements?

Reply:
Unfortunately `--cores` does not specify the actual cores usage. In this case it uses 2 cores to read the input, 2 cores to write to the output, and 2 cores for Cutadapt + 2 additional cores for Cutadapt + 1 core for Trim Galore itself. I can extend the comment with something like `# Actual core usage is 9, for more details see Trim Galore user guide` to provide a bit more info.